enhancements/ingress/route-secret-injection-for-external-certificate-management.md

```diff
@@ -0,0 +1,423 @@
+
+// RouteCertificateCondition contains details of the certificate being served by routers associated with
+// this route.
+type RouteCertificateCondition struct {
+	// Type is the type of the condition.
+    // Possible values include Default, Custom and Managed.
+	Type RouteCertificateConditionType `json:"type" protobuf:"bytes,1,opt,name=type,casttype=RouteCertificateConditionType"`
+	// Status is the status of the condition.
+	// Can be True, False, Unknown.
+	Status corev1.ConditionStatus `json:"status" protobuf:"bytes,2,opt,name=status,casttype=k8s.io/api/core/v1.ConditionStatus"`
+	// (brief) reason for the condition's last transition, and is usually a machine and human
+	// readable constant
+	Reason string `json:"reason,omitempty" protobuf:"bytes,3,opt,name=reason"`
+	// Human readable message indicating details about last transition.
+	Message string `json:"message,omitempty" protobuf:"bytes,4,opt,name=message"`
+	// RFC 3339 date and time when this condition last transitioned
+	LastTransitionTime *metav1.Time `json:"lastTransitionTime,omitempty" protobuf:"bytes,5,opt,name=lastTransitionTime"`
+}
+
+const (
+    // DefaultCertificate denotes that the user has not provided
+    // any custom certificate and the default generated certificate is being
+    // served on the route.
+    DefaultCertificate RouteCertificateConditionType = "Default"
+
+    // CustomCertificate denotes that there is a custom certificate
+    // being served on the route.
+    CustomCertificate RouteCertificateConditionType = "Custom"
+
+    // ManagedCertificate denotes that the annotation is in use to
+    // inject a third-party managed secret containing the certificate that
+    // is being served on the route.
+    ManagedCertificate RouteCertificateConditionType = "Managed"
+)
+
+```
+
+#### Variation
+
+##### Alternative to the tls-secret-reference annotation
+
+**_Note_**: This idea has been dropped in favor of a secret reference through an annotation.
+
+The alternative to the annotation was to provide a new API field in the Route API,
+through which the user could provide the certificate reference.
+
+```go
+type CertificateReference struct {
+	// certificate resource name.
+	// +optional
+	Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
+}
+
+Type RouteSpec struct {
+	// ...
+
+	CertificateRef CertificateReference `json:"certificateRef,omitempty" protobuf:"bytes,5,opt,name=certificateRef"`
+}
+```
+
+The reasoning for the field was mainly for future proofing the API to handle integrations
+with other third-party certificate management systems. But this API introduces inconsistency
+between how Ingresses are supported by cert-manager via `.spec.tls.secretName` and when using
+Routes, the user would have to provide the Certificate CR reference.
+
+This also adds coupling with external third-party API(s) and limits future enhancement to be
+a more generic solution for integrating with third-party certificate management systems.
```

Comment:
Does it? Are we not just saying that any certificate that can be provided by a secret can be used? That seem like it has no tighter coupling than the annotation API above

I actually don't see how this comment cannot apply to both the secret reference and annotation APIs, they both carry the same data, just one has a schema and the other doesn't

Reply:
Clarified [here](https://github.com/openshift/enhancements/pull/1307#discussion_r1085455475).